PSPs/psp-fungible-asset-standard.md

```diff
@@ -0,0 +1,318 @@
+
+## Specification
+
+A fungible asset is mainly made on two parts: Descriptions and Operations.
+
+### Asset Descriptions
+
+Asset descriptions offers a universal way to describe, display the details about a asset. This information may not be stored on-chain to reduce usage of on-chain storages. A centralized or decentralized asset registry may be used to discovery the asset description, which is out of the scope here.
+
+### Asset Operations
+
+An Asset implementation should provide set of operations to achieve common functionalities such as transfer and query balances.
+
+Write operations can be dispatched via extrinsic (i.e. transaction), smart contract call (on a chain with contracts module), or runtime module call (via public Rust methods). Query operations can be dispatched via RPC (`state_storage`), smart contract call, or runtime module call.
+
+## Specification
+
+### Asset Descriptions
+
+With TypeScript syntax:
+
+```typescript
+interface AssetDescription {
+  name: string;
+  symbol: string;
+  decimals: number;
+  uri?: string; // Substrate URI. See psp-uri
+  description?: string;
+  icon?: string; // HTTP URL or IPFS hash or dataURI
+}
+```
+
+Asset description can also be embedded in genesis file as additional properties. It is expected to be immutable but maybe changed under special circumstances (e.g. project rename / rebrand).
+
+### Asset Operations
+
+#### Data Types
+
+With TypeScript syntax:
+
+```typescript
+interface AssetOperationDescription {
+  types: TypeRegistry;
+  queries: {
+    [name: string]: {
+      parameterTypes?: Record<string, TypeWithHahser>;
+      returnType: string;
+      path: Path;
+    };
+  };
+  operations: {
+    [name: string]: {
+      parameterTypes: Record<string, TypeName>;
+      path: Path;
+    };
+  };
+  events: {
+  	[name: string]: {
+      types: Record<string, TypeName>;
+      path: Path;
+    } 
+  };
+}
+
+interface TypeWithHasher {
+  type: TypeName;
+  hasher: Hasher;
+}
+type TypeName = string;
+type Hasher = 'Blake2_128' | 'Blake2_256' | 'Blake2_128Concat' | 'Twox128' | 'Twox256' | 'Twox64Concat' | 'Identity';
+type Path = string | [ModuleName, MethodName, ...string[]]
+```
+
+Asset operation description stores the necessary information to perform asset operations and might need to be updated with new runtime versions. There should be a way to fetch the corresponding `AssetOperationDescription` . e.g. From a registry or a commonly known location.
+
+Note: Substrate Runtime Metadata is required to encode / decode dispatchable calls and events. The details of Substrate Runtime Metadata is out of scope.
+
+#### TypeRegistry
+
+Type registry stores the necessary type information to interact with the runtime. This is inspired by polkadot.js type registry.
+
+```typescript
+interface TypeRegistry {
+  [typename: string]: TypeDefination;
```

Comment:
I assume you meant to say `TypeDefinition`?

Reply:
oops, apparently I am really bad at spelling...